### PR TITLE
UICommon: Add foundation for future Steam API integrations

### PR DESCRIPTION
**Pull request by OatmealDome**
This PR adds the latest version of the Steamworks SDK to Externals, adds a new CMake option for creating a build with Steam integration, and adds a barebones class in UICommon that initializes the APIs. In addition, DolphinQt will quit on Steam builds if not running in a Steam context (i.e. the executable was ran directly) and the Steam client will relaunch it properly.

Dolphin does not actually do anything with the Steam APIs yet. I am submitting this in advance because of the issues discussed in the sections below.

**TODO**: Windows is not yet supported.

## Potential GPL Violation

This PR causes Dolphin to dynamically link to the Steamworks SDK, which is proprietary. The FSF appears to consider this to be a violation of the GPL. Note that by publishing to Steam, we are already in violation by this interpretation because the Steam client injects additional libraries into the app to power things like the Steam overlay.

This has been discussed in IRC around February and I *believe* the consensus was that this specific case isn't very important compared to what the GPL was intended to prevent (that is, proprietary code taking in GPL code), and it's probably fine to ignore. In addition, the "system libraries" exception may apply here, but the definition of what is a considered a "system library" is vague.

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
Since this includes binaries, can you make the SDK in the Exsternals a submodule instead?

**Comment by OatmealDome** [comment]
Valve doesn't host a git repo with the SDK contents, so I'd have to create one myself. Would it be OK to host it under my personal account or should it be under the dolphin-emu organization (I don't think I can create repos, so someone else would need to do that for me)?

**Comment by lioncash** [comment]
If we go the submodule route, it would be best to have it on the dolphin-emu org, so all org members can make changes to it if necessary

File: Source/Core/UICommon/SteamIntegration.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <functional>
```
Comment (lioncash) [inline Source/Core/UICommon/SteamIntegration.h:6]: These don't seem to be necessary in the header file

File: Source/Core/UICommon/SteamIntegration.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+#include <string>
+
+namespace Steam
```
Comment (lioncash) [inline Source/Core/UICommon/SteamIntegration.h:9]:
```suggestion
namespace UICommon::Steam
```

File: Source/Core/UICommon/SteamIntegration.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "UICommon/SteamIntegration.h"
+
+#ifdef STEAM
+
+#include "Common/FileUtil.h"
+#include "Common/MsgHandler.h"
+
+#include "steam/steam_api.h"
```
Comment (lioncash) [inline Source/Core/UICommon/SteamIntegration.cpp:11]: Since this is an external, it should be encased in `<>` rather than quotes.

**Comment by OatmealDome** [comment]
@lioncash that would be great, thank you!

**Comment by JosJuice** [comment]
> This PR causes Dolphin to dynamically link to the Steamworks SDK, which is proprietary. The FSF appears to consider this to be a violation of the GPL. Note that by publishing to Steam, we are already in violation by this interpretation because the Steam client injects additional libraries into the app to power things like the Steam overlay.
>
> This has been discussed in IRC around February and I *believe* the consensus was that this specific case isn't very important compared to what the GPL was intended to prevent (that is, proprietary code taking in GPL code), and it's probably fine to ignore. In addition, the "system libraries" exception may apply here, but the definition of what is a considered a "system library" is vague.

I would be careful here. Perhaps it could count as a system library for SteamOS, but for other OSes I don't really see how it would. And the only defense other than that seems to be if you consider dynamically linking to not be a GPL violation, which I don't think is a safe assumption to make. And if we're unsure if something is legal, I don't think we should do it.

**Comment by JosJuice** [comment]
After some studies of the GPL, it seems like GPLv3 is clearer than GPLv2 on this. GPLv3 says that "Corresponding Source" includes "the source code for shared libraries and dynamically linked subprograms that the work is specifically designed to require" (unless the system library exception applies). As I understand it, this would cover the case of Dolphin including the Steamworks library and making calls to it, but not the case of Steam's overlay hooking into Dolphin (as it's not designed to require Dolphin specifically). But IANAL.

GPLv2 could perhaps be read as allowing this type of dynamic linking but I wouldn't want to count on that without professional input.

**Comment by OatmealDome** [comment]
Forgot to mention... There is an alternative that we could potentially do - have some sort of shim in a separate process that links with Steamworks SDK service for us, and communicate with it over IPC. I think that puts our GPL code at a sufficient distance away from the prorpietary stuff. (Though, of course, this does introduce added complexity...)

We could also just drop Steam integration entirely. But I think Steam users expect at least some level of integration for anything released on there.

Not sure where we could get some sort of professional input on this.

**Comment by OatmealDome** [comment]
I will replace this PR with a more GPL-friendly solution soon-ish.

**Pull request closed without merging**